pkg/pillar/cmd/zedagent/hardwareinfo.go

```diff
@@ -74,14 +74,23 @@ func PublishHardwareInfoToZedCloud(ctx *zedagentContext, dest destinationBitset)
 		stDiskInfo.Model = *proto.String(disk.ModelNumber)
 		stDiskInfo.Wwn = *proto.String(fmt.Sprintf("%x", disk.Wwn))
 
-		attrSmart := new(info.SmartMetric)
-		attrSmart.ReallocatedSectorCt = getSmartAttr(types.SmartAttrIDRealLocatedSectorCt, disk.SmartAttrs)
-		attrSmart.PowerOnHours = getSmartAttr(types.SmartAttrIDPowerOnHours, disk.SmartAttrs)
-		attrSmart.PowerCycleCount = getSmartAttr(types.SmartAttrIDPowerCycleCount, disk.SmartAttrs)
-		attrSmart.ReallocatedEventCount = getSmartAttr(types.SmartAttrIDRealLocatedEventCount, disk.SmartAttrs)
-		attrSmart.CurrentPendingSector = getSmartAttr(types.SmartAttrIDCurrentPendingSectorCt, disk.SmartAttrs)
-		attrSmart.Temperature = getSmartAttr(types.SmartAttrIDTemperatureCelsius, disk.SmartAttrs)
-		stDiskInfo.SmartData = append(stDiskInfo.SmartData, attrSmart)
```

Comment:
The change you propose will work. I did it in the following way:

`stDiskInfo.SmartAttr = getSmartAttr(disk.SmartAttrs)`

and

```
func getSmartAttr(diskData []*types.DAttrTable) []*info.SmartAttr {
	attrResults := []*info.SmartAttr{} // Store pointers instead of structs
	
	for _, attr := range diskData {
		attrResult := &info.SmartAttr{ // Allocate on heap
			Id:            uint32(attr.ID),
			AttributeName: attr.AttributeName,
			RawValue:      uint64(attr.RawValue),
			Worst:         uint64(attr.Worst),
			Value:         uint64(attr.Value),
			Type:          attr.Type,
		}

		attrResults = append(attrResults, attrResult) // Append pointer
	}

	return attrResults
}
```

Reply:
I updated it to your nicer solution.